web/interceptor/tokens_test.go

```diff
@@ -0,0 +1,159 @@
+package interceptor_test
+
+import (
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+	"github.com/quickfeed/quickfeed/ci"
+	"github.com/quickfeed/quickfeed/internal/qtest"
+	"github.com/quickfeed/quickfeed/qf"
+	"github.com/quickfeed/quickfeed/qlog"
+	"github.com/quickfeed/quickfeed/scm"
+	"github.com/quickfeed/quickfeed/web"
+	"github.com/quickfeed/quickfeed/web/auth"
+	"github.com/quickfeed/quickfeed/web/interceptor"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func TestRefreshTokens(t *testing.T) {
+	db, cleanup := qtest.TestDB(t)
+	defer cleanup()
+	logger := qlog.Logger(t)
+	ags := web.NewQuickFeedService(logger.Desugar(), db, scm.TestSCMManager(), web.BaseHookOptions{}, &ci.Local{})
+
+	tm, err := auth.NewTokenManager(db, "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	lis := bufconn.Listen(BufSize)
+	bufDialer := func(context.Context, string) (net.Conn, error) {
+		return lis.Dial()
+	}
+	opt := grpc.ChainUnaryInterceptor(
+		interceptor.UnaryUserVerifier(logger, tm),
+		interceptor.TokenRefresher(logger, tm),
+	)
+	s := grpc.NewServer(opt) // skipcq: GO-S0902
+	qf.RegisterQuickFeedServiceServer(s, ags)
+
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			t.Errorf("Server exited with error: %v", err)
+			return
+		}
+	}()
+
+	admin := qtest.CreateFakeUser(t, db, 1)
+	user := qtest.CreateFakeUser(t, db, 56)
+
+	ctx := context.Background()
+	conn, err := grpc.DialContext(ctx, "bufnet", grpc.WithContextDialer(bufDialer), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to dial bufnet: %v", err)
+	}
+	defer conn.Close()
+
+	client := qf.NewQuickFeedServiceClient(conn)
+
+	adminCookie, err := tm.NewAuthCookie(admin.ID)
```

Comment:
Consider if these can be replaced with a helper along the lines to what I did in `access_control_test.go`:
```go
	f := func(t *testing.T, id uint64) context.Context {
		token, err := tm.NewAuthCookie(id)
		if err != nil {
			t.Fatal(err)
		}
		return qtest.WithAuthCookie(ctx, interceptor.AuthTokenString(token.Value))
	}
```
That is, it might be better to make a `qtest` func that does the same; it would need `ctx, t, tm, id` as arguments... so maybe it is easier to just do like above...